### ISSUE TITLE
POC: Improve deserialization performance

### ISSUE THREAD

**Issue by Tomas-Kraus** [eclipse-ee4j/yasson#335]
I implemented small POC to check whether there is some space for performance enhancement of deserialization algorithms.
New deserializer is available in org.eclipse.yasson.nternal.unmarshalling package. It's implemented as deterministic pushdown (stack) automaton parsing LL(1) grammar.
Only JSON Object to HashMap deserialization is implemented now.

I made 2 commits in my fork: https://github.com/Tomas-Kraus/yasson/tree/new_desearialization
 * 1st working POC, 20% performance gain against old parser
Probably better user readabler version with several redundant steps
* 2nd working POC, 80% performance gain against old parser
made few more optimizations and got quite nice results from JMH benchmark:

  - `NewDeserializerTest.testNew  thrpt    5  109.423 ± 0.567  ops/ms`
  - `NewDeserializerTest.testOld  thrpt    5   60.337 ± 0.555  ops/ms`

With new algorithm it did 80% more ops/ms when deserializing JSON Object with 50 elements to HashMap.
Preformance for larger data seems to be even few more percent better.

When fully implemented, it will be less effective than current POC, but I'm still expecting better performance than current code.
But there are also some drawbacks:
1. We have to rewrite all default deserializers
2. Only whole deserialization process is accessible trough API interface
3. Plugging of custom deserializers trough current API is possible, but quite challenging task

I posted it here because we have to make significant changes into serialization/deserialization code anyway when implementing Polymorphism. So it make sense to try to improve it's performance too. The only question is: How?
I would like to present one of the options to go. Another one is to try to improve current cose as much as possible.

**Comment by Tomas-Kraus** [comment]
@aguibert @Verdent @m0mus What do you guys think about this? :)

**Comment by m0mus** [comment]
PR to check:
https://github.com/Tomas-Kraus/yasson/pull/1

**Comment by mkarg** [comment]
@Tomas-Kraus I always apprciate 20-80% performance gain. As I am not a contributor to Yasson, I leave the technical evaluation to those who are. :-)

**Comment by aguibert** [comment]
great news @Tomas-Kraus, looking forward to evaluating it.  It is kind of hard to view what the proposed changes are because the PR https://github.com/Tomas-Kraus/yasson/pull/1/files includes many commits that we have already merged into master and are unrelated to the POC.

Can you please rebase what you have on the current yasson/master branch and make a "draft PR" (newish github feature) on the Yasson repo? It will be easier to view that way and since it is a draft nobody can accidentally merge it until you mark it as a "regular" PR

**Comment by Tomas-Kraus** [comment]
Looks like git was a bit confused. I did rebase/forcepush so code changes are visible now.
Also, this is far from being finished, it's just another concept of deserializers implementation and I want to know whether you like it or not. :)

**Comment by aguibert** [comment]
thanks, now I can review the changes much more easily. I've added a comment on the PR on your personal repo, but I would suggest creating a "Draft PR" on the main Yasson repo so we can potentially have wider participation

**Comment by aguibert** [comment]
If it improves performance and simplifies the codebase, I'm all for it. Main question will be backwards compatibility but I think we can rely on our unit tests and TCK tests here.

**Comment by Tomas-Kraus** [comment]
Yes, it must pass all jUnits and TCK. :)

**Comment by nimo23** [comment]
According to https://github.com/fabienrenaud/java-json-benchmark, performance of yasson in compare to jackson is not really good.

**Comment by aguibert** [comment]
I don't see anywhere on the java-json-benchmark where it actually mentions Yasson. The only related thing it only compares the reference implementation of JSON-P.

**Comment by nimo23** [comment]
I don't see anywhere on the java-json-benchmark where it actually mentions Yasson. 

@aguibert yasson is there: look at the chart on the X axis (for example, the first chart on the 3rd position from the left).

**Comment by nimo23** [comment]
> I don't see anywhere on the java-json-benchmark where it actually mentions Yasson. 

@aguibert yasson is evaluated in this benchmark: look at the chart on the X axis (for example, the first chart ("Deserialization performance") on the 3rd position from the left).

Also look at the benchmarked providers here: https://github.com/fabienrenaud/java-json-benchmark/blob/master/src/main/java/com/github/fabienrenaud/jjb/support/BenchSupport.java on line 18.

Why does `jackson` or `jsonDsl` has so much better performance in compare to `yasson`?

**Comment by aguibert** [comment]
Ah, I see Yasson is mentioned on the graph but it doesn't say what version they used. Never heard of dsljson (or anyone using it), but Jackson has been around for many years longer than JSON-B/Yasson and therefore has had more time to optimize performance. Our primary focus has been on adding functionality, rather than performance.

Yasson is still in the 1.0 version, and it doesn't surprise me that Jackson performs better. As you've found on this issue, @Tomas-Kraus is working on improving performance.

In any case, I don't think JSON databinding performance makes that much difference in the overall picture of application performance. I've ran tests in the past that compare a JAX-RS application using JSON-B vs. Jackson, and once you add in network I/O then JSON databind performance becomes negligible. If I were a user trying to evaluate which JSON library to choose, I would look at factors like adoption, compatibility, and stability before I looked at performance.

**Comment by nimo23** [comment]
>Ah, I see Yasson is mentioned on the graph but it doesn't say what version they used. 

According to https://github.com/fabienrenaud/java-json-benchmark/blob/master/build.gradle it uses v 1.0.1.

> If I were a user trying to evaluate which JSON library to choose, I would look at factors like adoption, compatibility, and stability before I looked at performance.

Definitly. Well, performance and memory consumption is still important. If I were a maintainer, I would look especially to Jackson or dsljson what techniques they use to have them so significantly increased performance. But I must admit I have never performance issues with yasson..

I mentioned the java-json-benchmark only FYI:)

**Comment by chrisgleissner** [comment]
In case anyone is interested in an up-to-date view on this topic, comparing the performance of the latest versions of Jackson and Yasson on an up-to-date JDK, I've updated the aforementioned java-json-benchmark via a fork. This uses the latest released versions of both frameworks as of today, all evaluated on Amazon Coretto JDK 11.0.5. You can find the results in the README at https://github.com/chrisgleissner/java-json-benchmark I also included the command-line invocations I used for easy reproduction of the test. 

In a nutshell, Jackson serialization is 2.12 as fast as Yasson. Deserialization is 3.8 times as fast. Jackson-Afterburner is even faster.

**Comment by Tomas-Kraus** [comment]
@chrisgleissner can you run it with original yasson. I would like to see numbers of original and this yasson.

**Comment by aguibert** [comment]
thanks for creating that comparison @chrisgleissner

@Tomas-Kraus you can run it on your own with the following steps:
```
git clone git@github.com:chrisgleissner/java-json-benchmark.git
cd java-json-benchmark
./run ser --apis databind --libs yasson,jackson --datatype users
```
To change the version of Yasson used you can edit the build.gradle file in the root of the repository

**Comment by nimo23** [comment]
@aguibert does yasson uses `java.lang.reflect` under the hood? If so, then please consider to use `java.lang.invoke.LambdaMetafactory`. It's way faster.

**Comment by Tomas-Kraus** [comment]
Well, looks like I found a dead end for JsonValue processing. :D
```
Benchmark                                 Mode  Cnt    Score    Error   Units
NewJsonValueDeserializationTest.testNew  thrpt    5    2.504 ±  0.023  ops/ms
NewJsonValueDeserializationTest.testOld  thrpt    5  227.769 ± 14.534  ops/ms
```
So using Json.createArrayBuilder() and build value is **200** times slower than old code using parser.getArray().

Looks like this part of old deserializer was quite good and I won't improve it much.

Current progress:
**[ERROR] Tests run: 440, Failures: 27, Errors: 58, Skipped: 1**

**Comment by chrisgleissner** [comment]
Please note the the maintainer of the Json benchmark project that I cloned for my benchmark runs has now updated his results using Yasson 1.0.6. He also upgraded other frameworks. The benchmark uses JMH and you can find it at https://github.com/fabienrenaud/java-json-benchmark
